File: 백준/Silver/15649. N과 M （1）/N과 M （1）.py

def solve(index,arr):

    # 1. 기저조건 index == M : M의 자리까지 숫자가 모여야함
    if index == M:
        result.append(arr)
        return
    # 2. 구현
    # 1부터 N까지 반복문 돌면서 방문 안한 숫자를 추가하고 재귀 돌기
    for i in range(N):
        if not visited[i+1]:
            visited[i+1] = 1
            solve(index+1,arr+[i+1])
            # 끝까지 다 돌고 나오면 직전에 방문했던 표시는 삭제
            visited[i+1]=0

N, M = map(int,input().split())
visited = [0] * (N+1)

result = []
solve(0,[])
for final in result:
    print(*final)

File: 백준/Silver/15649. N과 M （1）/N과 M （1）.java
import java.io.*;
import java.util.ArrayList;

public class Main {
    static int N, M;
    static boolean[] visited;
    static ArrayList<Integer> numbers = new ArrayList<>();
    static StringBuilder  sb = new StringBuilder();

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        // 입력 받기
        String[] input = br.readLine().split(" ");
        N = Integer.parseInt(input[0]);
        M = Integer.parseInt(input[1]);

        visited = new boolean[N + 1]; // 1부터 N까지 사용

        backtrack(0);

        System.out.println(sb);
    }

    private static void backtrack(int depth) throws IOException {
        // M개의 숫자를 모두 선택한 경우
        if (depth == M) {
        for(int num : numbers){
            sb.append(num).append(' ');
        }
       sb.append('\n');
        return;
    }

        // 1부터 N까지 숫자를 선택
        for (int i = 1; i <= N; i++) {
            if (!visited[i]) { // 방문하지 않은 숫자만 선택
                visited[i] = true;
                numbers.add(i); // ArrayList에 숫자를 추가
                backtrack(depth + 1); // 다음 숫자를 선택
                numbers.remove(numbers.size() - 1); // 백트래킹 - 마지막 숫자를 제거
                visited[i] = false; // 방문 상태를 초기화
            }
        }
    }
}
